proto/accessgraph/v1alpha/query.proto

```diff
@@ -0,0 +1,125 @@
+// Copyright 2023 Gravitational, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+syntax = "proto3";
+
+package accessgraph.v1alpha;
+
+import "teleport/legacy/client/proto/event.proto";
+import "teleport/legacy/types/types.proto";
+
+option go_package = "github.com/gravitational/teleport/gen/proto/go/accessgraph/v1alpha;accessgraphv1alpha";
+
+// AccessGraphService is a service for interacting the access graph service.
+service AccessGraphService {
+  // Query queries the access graph.
+  // Currently only used by WebUI.
+  rpc Query(QueryRequest) returns (QueryResponse);
+  // GetFile gets a static UI file from the access graph container.
+  rpc GetFile(GetFileRequest) returns (GetFileResponse);
+
+  // SendEvent sends an event to the access graph service.
+  rpc SendEvent(SendEventRequest) returns (SendEventResponse);
+
+  // SendResource sends a resource to the access graph service.
+  rpc SendResource(SendResourceRequest) returns (SendResourceResponse);
+}
+
+// Node is a node in the access graph.
+message Node {
+  // ID is the unique ID of the node.
+  string id = 1;
+  // Kind is the kind of the node, ex: "user", "user_group", "resource", etc.
+  string kind = 2;
+  // SubKind is the subkind of the node.
+  string sub_kind = 3;
+  // Name is the name of the node.
+  string name = 4;
+  // Labels is a set of labels for the resource.
+  map<string, string> labels = 5;
+  // Hostname is the hostname of the node.
+  string hostname = 6;
+  // Properties is a set of properties for the resource.
+  map<string, string> properties = 7;
+}
+
+// Edge is an edge in the access graph.
+message Edge {
+  // from is the ID of the node the edge is from.
+  string from = 1;
+  // to is the ID of the node the edge is to.
+  string to = 2;
+  // type is the type of the edge, e.g. "member_of", "belongs_to", etc.
+  string type = 3;
+}
+
+// QueryRequest is a request to query the access graph.
+message QueryRequest {
+  // query is a SQL query.
+  string query = 1;
+}
+
+// QueryResponse is a response to a query.
+message QueryResponse {
+  repeated Node nodes = 1;
+  repeated Edge edges = 2;
```

Comment:
Would representing outbound edges in the Node itself make this simpler? Conceptually I would expect nodes to have edges, not edges "hanging" by themselves.

Reply:
An edge is always from something to something else, right? So there's always a source and destination node? If that's true, then the design I'm suggesting would work.

Reply:
>An edge is always from something to something else, right?

Yes

>So there's always a source and destination node?

Yes, but one query can return only a part of a graph. In this case the client need to combine two or more different result into a single graph. If the edges are separate, you can easily create a map that holds `to/from` edge and merge all nodes: ex:
```json
{
    "nodes": { "name": "node1"},
    "edges": { "from": "node1", "to": "node2"}
 }
 ```
 
 second response:
 ```json
{
    "nodes" { "name": "node2"},
 }
 ```
 IMO it's easier if everything is separate. 
 > If that's true, then the design I'm suggesting would work.
 
Agree with the principle. It works in both cases.

Reply:
Sure.

> one query can return only a part of a graph

I had no clue this could happen.